### PR TITLE
[MRG] add abundance-weighted columns to gather output

### LINKED ISSUE #2240: report abundance weighted hashes/bp from gather and tax?
In #2239, I found myself wanting the a summarized weighted bp -- the total bp matched via sourmash gather. We don't currently summarize that in `sourmash tax`, partially because we don't report it from gather. 

- Is it valid to just multiply `f_unique_weighted` by the `query_scaled`?
- How can we get at the weighted unclassified bp? afaict, we don't currently store the original `total_weighted_bp` for the query...

Backing up to `sourmash gather`, we could:
- Report `weighted_unique_intersect_bp` from `gather`.
- Report query weighted bp (to allow weighted unclassified)

These would allow summarization in tax.

Somewhat separately,  @taylorreiter has suggested that reporting hashes, rather than bp, would be more useful so as not to inflate our numbers / impact downstream methods, e.g. diff abundance.

She suggested some/any of the following:
```
n_hashes
n_hashes_weighted
n_unique_hashes
n_unique_hashes_weighted
```

But in thinking through reporting hashes, I realized this creates a secondary issue. Since gather can function at multiple scaled vals and will dynamically downsample as needed, `n_hashes` matched in one comparison may _mean_ something different than `n_hashes` matched in the another, if the comparison scaled was different. Only an issue for matching against db's with mixed scaled values.

Just documenting rn, thoughts welcome.

### PR DESCRIPTION
**Pull request by ctb**
This PR adds three abundance-weighted columns to gather output:
* `n_unique_weighted_found` - the summed abundances of the found hashes at each step
* `sum_weighted_found` - the running total of `n_unique_weighted_found`, cumulative at this step
* `total_weighted_hashes` - the sum total of all hash abundances

It also updates the `kreport` format of `sourmash tax metagenome` to output weighted bp estimates.

Fixes #2240.

## TODO

- [x] add loading and error-message-if-missing to sourmash tax
- [x] update `kreport` format from #2239 to support weighted reporting
- [ ] implement & test what happens to these values with `--ignore-abundance`
- [x] clean up redundant `total_abund` in `GatherResult`
- [x] revisit `GatherResult` and `FracMinHashComparison` classes for other cleanup
- [x] consider building more robust/bigger tests of the weighting 🙄

### CODE REVIEW

**Comment by ctb** [comment]
@bluegenes ok, updated. On inspection I think that perhaps it is/was possible to calculate some or all of this stuff before from the CSV output - test `test_gather_abund_10_1` in `test_sourmash.py` seems to do so, at any rate! - but I like the new column names.

If you get a chance to look at this, would appreciate your interim stamp of approval :). I might not get to work on this more until later today.

**Comment by bluegenes** [comment]
> @bluegenes ok, updated. On inspection I think that perhaps it is/was possible to calculate some or all of this stuff before from the CSV output - test `test_gather_abund_10_1` in `test_sourmash.py` seems to do so, at any rate! - but I like the new column names.
> 
> If you get a chance to look at this, would appreciate your interim stamp of approval :). I might not get to work on this more until later today.

STAMP --the new columns are great and I think they'll be really helpful!

ye, re already able to calculate most things - the main issue was needing total_weighted_missed for the unclassified portion, which I wasn't sure how to calculate without outputting the total weighted query hashes from gather.

**Comment by ctb** [comment]
@bluegenes I think I'm missing one thing before I can properly finish this PR off - I need to regenerate `tests/test-data/tax/test1.gather.csv`, but I don't have `test1.sig` handy.  Do you know where this is?

**Comment by bluegenes** [comment]
> @bluegenes I think I'm missing one thing before I can properly finish this PR off - I need to regenerate `tests/test-data/tax/test1.gather.csv`, but I don't have `test1.sig` handy. Do you know where this is?

I believe it was this one, just renamed: https://github.com/taylorreiter/2021-sourmash-taxonomy-hackathon/tree/main/outputs/sigs

..realizing now I probably should have added it, sry!

`test1.gather.csv` is this file, renamed: https://github.com/taylorreiter/2021-sourmash-taxonomy-hackathon/blob/main/outputs/gather/HSMA33MX_gather_x_gtdbrs202_k31.csv

**Comment by ctb** [comment]
I think this is ready for review @bluegenes.

File: src/sourmash/tax/tax_utils.py [hunk 1]
```diff
@@ -296,7 +302,7 @@ def summarize_gather_at(rank, tax_assign, gather_results, *, skip_idents = [],
                     query_ani = containment_to_distance(fraction, ksize, scaled,
                                                         n_unique_kmers=qInfo.query_hashes, sequence_len_bp=qInfo.query_bp).ani
                 sres = SummarizedGatherResult(query_name, rank, fraction, lineage, query_md5,
-                                              query_filename, f_weighted_at_rank, bp_intersect_at_rank, query_ani)
+                                              query_filename, f_weighted_at_rank, bp_intersect_at_rank, query_ani, qInfo.total_weighted_hashes * scaled)
```
Comment (bluegenes) [inline src/sourmash/tax/tax_utils.py:305]: `scaled` is not in the gather results prior to v4.4 (added at he same time as `ksize`) -- might be an issue with earlier gather results

File: src/sourmash/tax/tax_utils.py [hunk 2]
```diff
@@ -488,7 +500,7 @@ def write_kreport(summarized_gather, csv_fp, *, sep='\t'):
             # total percent containment, weighted to include abundance info
             kresD['percent_containment'] = f'{res.f_weighted_at_rank:.2f}'
             # num bp contained. THIS IS NOT WEIGHTED... do we want to weight??
```
Comment (bluegenes) [inline src/sourmash/tax/tax_utils.py]:
```suggestion
            # weighted bp
```

File: src/sourmash/tax/tax_utils.py (hunk 1 above)
Comment (bluegenes) [reply src/sourmash/tax/tax_utils.py:305]: oh I see this is what you're checking for on like 478. Looks good!

Did we want to output the total weighted hashes in the csv summary too? Or just punt this until someone wants it?

**Review by bluegenes — APPROVED** [review]
looks good to me. thanks for adding!

File: src/sourmash/tax/tax_utils.py (hunk 1 above)
Comment (ctb) [reply src/sourmash/tax/tax_utils.py:305]: yep - handled elsewhere :)

**Pull request merged**